pyhap/characteristic.py

```diff
@@ -189,50 +151,44 @@ def override_properties(self, properties=None, valid_values=None):
 
     def notify(self):
-        """Notify clients about a value change. Sends the value.
-
-        .. note:: Non-blocking, i.e. does not wait for the update to be sent.
-        .. seealso:: accessory_driver.publish
-
-        :raise RuntimeError: When the broker is not set.
-        """
-        if not self.broker:
-            raise RuntimeError('Notify failed, because broker is not set')
-
-        data = {
-            'type_id': self.type_id,
-            'value': self.value,
+        """Notify clients about a value change. Sends the value."""
+        self.broker.publish(self.value, self)
+
+    def to_dict(self):
+        """Create a dict repr of this object for json serialization."""
+        json_dict = {
+            'iid': self.broker.iid_manager.get_iid(self),
+            'type': str(self.type_id).upper(),
+            'description': self.display_name,
+            'perms': self.properties['Permissions'],
+            'format': self.properties['Format'],
         }
-        self.broker.publish(data, self)
 
-    def to_HAP(self, iid_manager):
-        """Create a HAP representation of this Characteristic. Sends the value.
-
-        :param iid_manager: IID manager to query for this object's IID.
-        :type iid_manager: IIDManager
-
-        :return: A HAP representation.
-        :rtype: dict
-        """
-        hap_rep = {
-            "iid": iid_manager.get_iid(self),
-            "type": str(self.type_id).upper(),
-            "description": self.display_name,
-            "perms": self.properties["Permissions"],
-            "format": self.properties["Format"],
-        }
+        if self.properties['Format'] in HAP_FORMAT.NUMERIC:
+            json_dict.update({k: self.properties[k] for k in
+                              self.properties.keys() & _HAP_NUMERIC_FIELDS})
+        elif self.properties['Format'] == HAP_FORMAT.STRING:
+            if len(self.value) > 64:
+                json_dict['maxLen'] = min(len(self.value), 256)
+        if HAP_PERMISSIONS.READ in self.properties['Permissions']:
+            json_dict['value'] = self.value
 
-        if self.properties["Format"] in HAP_FORMAT.NUMERIC:
-            value_info = {k: self.properties[k] for k in
-                          self.properties.keys() & _HAP_NUMERIC_FIELDS}
-        else:
-            value_info = dict()
+        return json_dict
 
-        if self.properties["Format"] == HAP_FORMAT.STRING:
-            if len(self.value) > 64:
-                value_info["maxLen"] = min(len(self.value), 256)
-        if HAP_PERMISSIONS.READ in self.properties["Permissions"]:
-            value_info["value"] = self.value
+    def update_value(self, value):
+        """Called from broker for value change in Home app.
 
-        hap_rep.update(value_info)
-        return hap_rep
+        Change self.value to value and call callback.
+        The callback method must return a boolean that indicates if successful.
+        """
+        self.value = value
```

Comment:
Value should only be set after the `setter_callback` is successful, otherwise the `value` won't represent what the device did.